w3d_translator/read_xml.py

```diff
@@ -39,12 +47,54 @@ def read_xml(file):
     for tag in root.find("ParticleActionRoot") or []:
         pass  # Dict of ParticleActionList by name
 
-    # Globals
-    g = root.find("Global")
-    # TODO: Build <CameraPos> and <CaveCameraPos> (12)
-    # story["camera_pos"] = parse_camera_pos(g.find("CameraPos"))
-    # story["cave_camera_pos"] = parse_cave_camera_pos(g.find("CaveCameraPos"))
-    story["background"] = Color(g.find("Background"))
-    story["wand_navigation"] = WandNavigation(g.find("WandNavigation"))
-
     return story
+
+
+def parse_camera(xml: ET.Element) -> dict:
+    return {
+        "far_clip": float(xml.attrib["far-clip"]),
+        "placement": parse_placement(xml.find("Placement")),
+    }
+
+
+def parse_placement(xml: ET.Element) -> dict:
+    relative_to = xml.find("RelativeTo").text
+    position = str_to_tuple(xml.find("Position").text)
+    axis = None
+    look_at = None
+
+    axis_xml = xml.find("Axis")
+    if axis_xml is not None:
+        axis = {
+            "rotation": str_to_tuple(axis_xml.attrib["rotation"]),
+            "angle": float(axis_xml.attrib["angle"]),
```

Comment:
Totally agree here, looks like you could use it below too for the `look_at` dict

Reply:
Do you have any recommended reads on regex matching and that sort of thing in Python? Say for example running into a `(0.0, 0.0, -7.0)` type string (they're all over the xml) should always call `str_to_tuple`.

I did think/work through a recursive function to read the whole file at once but I found drilling through the objects for these more basic type conversions to be very tedious (the above would be something like `story["walls"]["Center"]["rotation"] = float(story["walls"]["Center]["rotation"])`.

Reply:
I'm also wondering if I should use asserts to validate the xml structure? No need to implement right now but given how unintuitive the xml already is (not to mention briefly reading through some of the other projects) I think it will pay off in the long run. Shall I add an issue?

Reply:
`import re; m = re.match("\(\s*-?(\d+(\.\d+)?),\s*-?(\d+(\.\d+)?),\s*-?(\d+(\.\d+)?)\)", "(0.0, 0.0, -7.0)")` should work some. The numbers are captured, but in groups 1, 3, and 5 (2, 4, and 6 are all the decimal portions of the numbers which are marked optional)

I'm into asserts! Throwing errors is a good thing

Reply:
I ended up checking for `bool`, `int`, `float`, and `tuple` using logic. Let me know if you think it can be tackled better/easier with regex though. I'm not super familiar with it so there'd be a learning curve for me first - definitely not opposed to learning

Reply:
Check that - switched `parse_string` to use regex. I'd appreciate a close set of eyes to make sure I set them up correctly 😄